### ISSUE TITLE
used-before-assignment error when using assignment operator inside a nested if comprehension

### ISSUE THREAD

**Issue by ozturkberkay** [PyCQA/pylint#8116]
### Bug description

```python
from dataclasses import dataclass
from typing import Optional


@dataclass
class C:
    val: Optional[str] = None


@dataclass
class B:
    val: Optional[C] = None


@dataclass
class A:
    val: B


def func() -> Optional[str]:
    ...


c = C()
b = B(val=c)
a = A(val=b)


res = func() or (val if (val := a.val.val) else "fallback")
```

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
************* Module test
test.py:29:17: E0601: Using variable 'val' before assignment (used-before-assignment)

-----------------------------------
Your code has been rated at 6.43/10
```

### Expected behavior

No `used-before-assignment` error. Removing `func() or` from the code gets rid of the error. Pylint should not complain regardless.

### Pylint version

```shell
pylint 2.16.0b1
astroid 2.13.3
Python 3.8.12 (default, Oct 12 2021, 06:23:56) 
[Clang 10.0.0 ]
```

### OS / Environment

macOS-10.16-x86_64-i386-64bit
